lib/dockerregistry/checks_test.go

```diff
@@ -168,3 +168,137 @@ func TestImageRemovalCheck(t *testing.T) {
 			fmt.Sprintf("checkError: test: %v imageRemovalCheck\n", test.name))
 	}
 }
+
+func TestImageSizeCheck(t *testing.T) {
+	srcRegName := reg.RegistryName("gcr.io/foo")
+	destRegName := reg.RegistryName("gcr.io/bar")
+	destRC := reg.RegistryContext{
+		Name:           destRegName,
+		ServiceAccount: "robot",
+	}
+	srcRC := reg.RegistryContext{
+		Name:           srcRegName,
+		ServiceAccount: "robot",
+		Src:            true,
+	}
+	registries := []reg.RegistryContext{destRC, srcRC}
+
+	image1 := reg.Image{
+		ImageName: "foo",
+		Dmap: reg.DigestTags{
+			"sha256:000": {"0.9"}}}
+	image2 := reg.Image{
+		ImageName: "bar",
+		Dmap: reg.DigestTags{
+			"sha256:111": {"0.9"}}}
+
+	const mbToByteShift = 20
+
+	var tests = []struct {
+		name       string
+		check      reg.ImageSizeCheck
+		manifests  []reg.Manifest
+		imageSizes map[reg.Digest]int
+		expected   error
+	}{
+		{
+			"Image size under the max size",
+			reg.ImageSizeCheck{
+				MaxImageSize:    1,
+				DigestImageSize: make(reg.DigestImageSize),
+			},
+			[]reg.Manifest{
+				{
+					Registries: registries,
+					Images: []reg.Image{
+						image1,
+					},
+					SrcRegistry: &srcRC},
+			},
+			map[reg.Digest]int{
+				"sha256:000": 1 << mbToByteShift,
+			},
+			nil,
+		},
+		{
+			"Image size over the max size",
+			reg.ImageSizeCheck{
+				MaxImageSize:    1,
+				DigestImageSize: make(reg.DigestImageSize),
+			},
+			[]reg.Manifest{
+				{
+					Registries: registries,
+					Images: []reg.Image{
+						image1,
+					},
+					SrcRegistry: &srcRC},
+			},
+			map[reg.Digest]int{
+				"sha256:000": 5 << mbToByteShift,
+			},
+			fmt.Errorf("The following images were over the max file size" +
+				" of 1MiB: foo (5 MiB)\n"),
+		},
+		{
+			"Multiple images over the max size",
+			reg.ImageSizeCheck{
+				MaxImageSize:    1,
+				DigestImageSize: make(reg.DigestImageSize),
+			},
+			[]reg.Manifest{
+				{
+					Registries: registries,
+					Images: []reg.Image{
+						image1,
+						image2,
+					},
+					SrcRegistry: &srcRC},
+			},
+			map[reg.Digest]int{
+				"sha256:000": 5 << mbToByteShift,
+				"sha256:111": 10 << mbToByteShift,
+			},
+			fmt.Errorf("The following images were over the max file size" +
+				" of 1MiB: bar (10 MiB), foo (5 MiB)\n"),
+		},
+		{
+			"Image sizes are <= 0",
+			reg.ImageSizeCheck{
+				MaxImageSize:    1,
+				DigestImageSize: make(reg.DigestImageSize),
+			},
+			[]reg.Manifest{
+				{
+					Registries: registries,
+					Images: []reg.Image{
+						image1,
+						image2,
+					},
+					SrcRegistry: &srcRC},
+			},
+			map[reg.Digest]int{
+				"sha256:000": 0,
+				"sha256:111": -5 << mbToByteShift,
+			},
+			fmt.Errorf("The following images had an invalid file size of 0" +
```

Comment:
It might be worth creating a custom error type. Here you're checking for the string representation of the error, which can get a bit unwieldy during comparisons because of how strings can be formatted. If you use a custom error type then you can compare the custom structs directly against each other.

Reply:
Done. I've added an ImageSizeError with an Error() function in checks.go